keps/sig-storage/2857-runtime-assisted-pv-mounts/README.md

```diff
@@ -0,0 +1,1767 @@
+post-mount configuration (specified by the pod):
+- The mutating webhook `fs-deferral-strategy-assigner.k8s.io` determines the pod
+spec and the file system deferral capabilities of the RuntimeClass and CSI
+plugin backing the PVCs associated with `fast-storage` are all compatible with
+respect deferring file operations. Therefore, it sets `FSDeferralStrategy` in
+both `PersistentVolumeClaimVolumeSource`s of the pod spec to `kata`.
+- The pod is scheduled to an appropriate node with the specified runtime
+installed while the PVCs gets bound to PVs either prior to the scheduling or
+right after scheduling depending on volume binding mode of the storage class.
+- The Kubelet invokes CSI `NodePublishVolume` (for both volumes on the pod)
+passing the following based on the pod spec: `fs_deferral_strategy` set to
+`kata`, `volume_supplemental_group` set to `100`, `volume_supplemental_group`
+set to `OnRootMismatch`.
+- The CSI plugin processes CSI NodePublishVolume. Upon finding a
+`fs_deferral_strategy` set to `kata`, it ensures Kata components on the host is
+indeed capable of handling file system mounts. Next, as part of the `kata`
+deferral strategy, the CSI plugin populates a metadata file with file system
+mount details that Kata expects (including block device name, target publish
+path on the host, file system type, mount options, fsGroup GID and
+fsGroupChangePolicy) and interacts with the Kata components on the host to
+ensure the file is registered by Kata. Note that the CSI plugin does not have
+access to selinux and subpath details of the pod as those are passed directly
+through CRI and OCI using existing mechanisms in Kubelet and CRI container
+runtimes respectively.
+- The Kubelet prepares the CRI parameters for the pod. Given
+`fs_deferral_strategy` is set, when preparing the mount specs, Kubelet does not
+probe the subpath or bind mount it to a separate path under `/proc/<kubelet
+pid>/...`. Instead, Kubelet simply concatenates the host path of the volume with
+the subpath specified for a container mount and passes it to CRI container
+runtime.
+- The CRI container runtime prepares OCI specs based on the CRI parameters.
+There are no changes in this step relative to what exists today.
+- The Kata runtime gets invoked with the OCI specs. Using the OCI mount specs,
+Kata runtime probes for (accounting for any subpaths) and detects the metadata
+file populated by the CSI plugin (as part of the `kata` deferral strategy). Kata
+detects the presence of subpaths upon finding the volume publish path is a
+prefix of the host paths in the mounts in the OCI specs (populated from CRI
+which in turn was originally populated by Kubelet as described above). Note that
+selinux labels are also specified in OCI spec but Kata does not support selinux
+enforcement today within the guest environment and therefore ignores the selinux
+details. The Kata runtime creates a micro VM for the pod and passes the block
```

Comment:
The OCI runtime handler already gets the selinux labels through OCI config.json for each container. So if a runtime has an environment where selinux is enforced and it wants to chcon files with the selinux label, it can (based on the selinux label specified in the config.json).